js/components/source-manager.js

```diff
@@ -0,0 +1,178 @@
+    var self = this;
+    var authApi = params.model.authApi;
+    self.config = config;
+    self.model = params.model;
+    self.loading = ko.observable(false);
+    self.dirtyFlag = self.model.currentSourceDirtyFlag;
+
+    self.selectedSource = params.model.currentSource;
+    self.selectedSourceKey = params.model.selectedSourceKey;
+
+    self.options = {};
+
+    self.isAuthenticated = authApi.isAuthenticated;
+
+    self.canReadSource = function() {
+      return (config.userAuthenticationEnabled && self.isAuthenticated()
+        && authApi.isPermittedReadSource(self.selectedSourceKey())) || !config.userAuthenticationEnabled
+        || !self.selectedSourceKey();
+    };
+
+    self.isDeletePermitted = function() {
+      return (config.userAuthenticationEnabled && self.isAuthenticated() &&
+        authApi.isPermittedDeleteSource(self.selectedSource().key())) || !config.userAuthenticationEnabled;
+    };
+
+    self.canEdit = function() {
+      return (config.userAuthenticationEnabled && self.isAuthenticated() &&
+        authApi.isPermittedEditSource(self.selectedSourceKey())) || !config.userAuthenticationEnabled;
+    };
+
+    self.canSave = function () {
+      console.log('canEdit', self.canEdit());
+      return (self.selectedSource() && self.selectedSource().name() && self.selectedSource().key()
+        && self.selectedSource().connectionString() && self.canEdit());
+    };
+
+    self.canDelete = function () {
+      return (self.selectedSource() && self.selectedSource().key() && self.isDeletePermitted());
+    };
+
+    self.options.dialectOptions = [
+      { name: 'PostgreSQL', id: 'postgresql' },
+      { name: 'SQL server', id: 'sqlserver' },
+      { name: 'Oracle', id: 'oracle' },
+      { name: 'Amazon Redshift', id: 'redshift' },
+      { name: 'Google BigQuery', id: 'bigquery' },
+      { name: 'Impala', id: 'impala' },
+      { name: 'Microsoft PDW', id: 'pdw' },
+      { name: 'IBM Netezza', id: 'netezza' },
+    ];
+
+    self.sourceCaption = ko.computed(function(){
+      return (self.model.currentSource() == null || self.model.currentSource().key() == null) ? 'New source' :
+        'Source ' + self.model.currentSource().name();
+    });
+
+    self.newSource = function () {
+      self.selectedSource(new Source());
+      self.dirtyFlag(new ohdsiUtil.dirtyFlag(self.selectedSource()));
+    };
+
+    self.save = function () {
+      var source = {
+        name: self.selectedSource().name() || null,
+        key: self.selectedSource().key() || null,
+        dialect: self.selectedSource().dialect() || null,
+        connectionString: self.selectedSource().connectionString() || null,
+        daimons: ko.toJS(self.selectedSource().daimons()).filter(function(d){ return d.enabled; }).map(function(d){
+          return lodash.omit(d, ['enabled']);
+        }),
+      };
+      sourceApi.saveSource(self.selectedSourceKey(), source).then(function(){
+        return sourceApi.initSourcesConfig();
```

Comment:
Is this call to init all of the sources necessary or could we process the new data source individually? In our environment with many data sources configured, the save seems to take a very long time due to this call. Alternatively, we could lock the screen while it is saving so that users cannot re-submit the same information.